src/unstable/read.rs

```diff
@@ -0,0 +1,751 @@
+
+pub(crate) fn find_entry_content_range<R>(
+    data: &ZipFileData,
+    mut reader: R,
+) -> Result<io::Take<R>, ZipError>
+where
+    R: Read + Seek,
+{
+    // TODO: use .get_or_try_init() once stabilized to provide a closure returning a Result!
+    let data_start = match data.data_start.get() {
+        Some(data_start) => *data_start,
+        None => find_data_start(data, &mut reader)?,
+    };
+
+    reader.seek(io::SeekFrom::Start(data_start))?;
+    Ok(reader.take(data.compressed_size))
+}
+
+pub(crate) fn construct_decompressing_reader<R>(
+    compression_method: &CompressionMethod,
+    reader: R,
+) -> Result<EntryReader<R>, ZipError>
+where
+    /* TODO: this really shouldn't be required upon construction (especially since the reader
+     * doesn't need to be mutable, indicating the Read capability isn't used), but multiple of our
+     * constituent constructors require it. We should be able to make upstream PRs to fix these. */
+    R: Read,
+{
+    match compression_method {
+        &CompressionMethod::Stored => Ok(EntryReader::Stored(reader)),
+        #[cfg(feature = "_deflate-any")]
+        &CompressionMethod::Deflated => {
+            let deflate_reader = DeflateDecoder::new(reader);
+            Ok(EntryReader::Deflated(deflate_reader))
+        }
+        #[cfg(feature = "deflate64")]
+        &CompressionMethod::Deflate64 => {
+            let deflate64_reader = Deflate64Decoder::new(reader);
+            Ok(EntryReader::Deflate64(deflate64_reader))
+        }
+        #[cfg(feature = "bzip2")]
+        &CompressionMethod::Bzip2 => {
+            let bzip2_reader = BzDecoder::new(reader);
+            Ok(EntryReader::Bzip2(bzip2_reader))
+        }
+        #[cfg(feature = "zstd")]
+        &CompressionMethod::Zstd => {
+            let zstd_reader = ZstdDecoder::new(reader).unwrap();
+            Ok(EntryReader::Zstd(zstd_reader))
+        }
+        #[cfg(feature = "lzma")]
+        &CompressionMethod::Lzma => {
+            let reader = LzmaDecoder::new(reader);
+            Ok(EntryReader::Lzma(Box::new(reader)))
+        }
+        #[cfg(feature = "xz")]
+        &CompressionMethod::Xz => {
+            let reader = XzDecoder::new(reader);
+            Ok(EntryReader::Xz(reader))
+        }
+        /* TODO: make this into its own EntryReadError error type! */
+        _ => Err(ZipError::UnsupportedArchive(
+            "Compression method not supported",
+        )),
+    }
+}
+
+pub(crate) enum CryptoReader<R> {
+    ZipCrypto(ZipCryptoReaderValid<R>),
+    #[cfg(feature = "aes-crypto")]
+    Aes(AesReaderValid<R>),
+}
+
+impl<R> Read for CryptoReader<R>
+where
+    R: Read,
+{
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            CryptoReader::ZipCrypto(r) => r.read(buf),
+            #[cfg(feature = "aes-crypto")]
+            CryptoReader::Aes(r) => r.read(buf),
+        }
+    }
+}
+
+pub(crate) enum CryptoVariant {
```

Comment:
Change this to a more descriptive name, maybe `CryptoKeyValidationSource`.

Reply:
Done! Thanks!